sycl/source/device.cpp

```diff
@@ -208,5 +208,78 @@ pi_native_handle device::getNative() const { return impl->getNative(); }
 
 bool device::has(aspect Aspect) const { return impl->has(Aspect); }
 
+void device::ext_oneapi_enable_peer_access(const device &peer) {
+  const RT::PiDevice Device = impl->getHandleRef();
+  const RT::PiDevice Peer = peer.impl->getHandleRef();
+  if (Device != Peer) {
+    auto Plugin = impl->getPlugin();
+    pi_result result =
+        Plugin->call_nocheck<detail::PiApiKind::piextEnablePeerAccess>(Device,
+                                                                       Peer);
+    if (result != PI_SUCCESS) {
+      char *message = nullptr;
+      Plugin->call<detail::PiApiKind::piPluginGetLastError>(&message);
+      throw sycl::exception(make_error_code(errc::runtime), message);
+    }
```

Comment:
Do we really need this special handling for errors?  `plugin::call()` should do that for `PI_ERROR_PLUGIN_SPECIFIC_ERROR`.

Reply:
Does this (https://github.com/intel/llvm/pull/8303#issuecomment-1549655540) answer your question?

Or maybe I misunderstand and you propose something simpler that achieves:

- native backend calls (e.g. L0/cu functions) always lead to native error messages being passed to what()

?

Reply:
Maybe the simplest way to think about it is by way of a specific example relevant to this extension: in the cuda backend `ext_oneapi_disable_peer_access` calls a `cuCtxDisablePeerAccess` which returns `CUDA_ERROR_PEER_ACCESS_NOT_ENABLED` if the user has not previously called `ext_oneapi_enable_peer_access`

Now in L0 backend the same situation will not lead to any error. `CUDA_ERROR_PEER_ACCESS_NOT_ENABLED` is an example of a type of error which only has real meaning for the cuda backend. However I'm pretty sure that there are in general for each backend lots of similar errors. So `piPluginGetLastError` is used in order to retrieve these errors messages so that they can be passed to the sycl::exception. The sycl::exception needs to be thrown according to the sycl spec. It is true that there are many cases in dpc++ where sycl::exceptions aren't thrown atm, but as I understand it this has to change according to the spec.

Reply:
Although I think it is still an open question for whether the problem has to be dealt with in the same way in every backend: I think it can and ideally every backend can make the same PI calls to get error message information.

I think the key think is to fix the problem we have at the moment where native error information is simply thrown away. Meaning that as a user there is either no error sign-posting in a lot of cases, or violation of the sycl spec (e.g. throwing cuResults directly and not sycl::exception).

Reply:
We did consider briefly with the UR people the possibility of adding lots of new PI variants corresponding to required backend errors: to maintain an effective 1:1 mapping between backend errors and PI errors. UR people were I think against the idea of expanding PI error codes at all. But generally I think it is bad because you either lose information or you would have the problem of an exploding number of PI errors to implement to keep a 1:1 mapping.

Reply:
Maybe that would be a reason for adding a new query for allowing backend-specific error-codes? It does blur some barriers.

Either way, it does not address my concerns about this. We are potentially hurting the user more than we're helping them here. What happens if a backend decides that it want to return a known PI error code here? There would be no error code to get, so it might be getting an old error code. Maybe we need to expand the handling of `PI_ERROR_PLUGIN_SPECIFIC_ERROR` in `checkPiResult` to be better at using the message in actual backend error cases and not just warning cases.

Also, as another note, I just noticed you can pass an errc to `plugin::call`, so you can control the error code too, so it seems to just be the better handling of backend-specific errors that we want.

Reply:
OK I see that I missed that `Plugin::check` now throws a `sycl::exception` directly if it does not return `PI_SUCCESS`, and this it is actually getting the last message in the `PI_ERROR_PLUGIN_SPECIFIC_ERROR` case. So what I added in 
```
    if (result != PI_SUCCESS) {
      char *message = nullptr;
      Plugin->call<detail::PiApiKind::piPluginGetLastError>(&message);
      throw sycl::exception(make_error_code(errc::runtime), message);
    }
```

is just a duplication. So I could go back to wrapping the e.g. `cuCtxDisablePeerAccess` call in `PI_CHECK_ERROR` then I think my question is whether it is OK to simply rely on a UR version of `https://github.com/intel/llvm/blob/sycl/sycl/plugins/cuda/pi_cuda.cpp#LL161C3-L161C3` to output a specific error message within the PI/UR? Or whether we should instead make sure that `map_error` in pi_cuda.cpp maps all CUDA errors to `PI_ERROR_PLUGIN_SPECIFIC_ERROR`, and calls `setErrorMessage(message, PI_ERROR_PLUGIN_SPECIFIC_ERROR);` prior to returning, then relies on `checkPiResult` to set the `what()` in the `sycl::exception`.

Reply:
So now I see that l0 is already doing what we were suggesting and I think our plan is adopt the same approach that you suggested which is basically think we were thinking and that I've now updated this PR to do for the P2P PI impl: basically:
call `Plugin::call`:


https://github.com/intel/llvm/pull/8303/files#diff-e5a5ec910bab410f31617da9dfb56ee9e435769e7d6a830e85094c07b80c5454R225

If a native api returns a native error then we set an appropriate last message using the native error (and return PI_ERROR_PLUGIN_SPECIFIC_ERROR):

https://github.com/intel/llvm/pull/8303/files#diff-7525901710934f7bdb2ad36238c4b67163f112d3bd233db7af0b0078b5b01e80R5866

 The only issue is that hip/cuda backends aren't using the correct mechanism currently: so what we need to do is remove map_error from cuda/hip and follow this pattern in future: This might just happen during the UR port.